source/grid/grid_generator.cc

```diff
@@ -6480,9 +6480,39 @@ namespace GridGenerator
     // Top (Z+) part of the cylinder has boundary id 3
 
     // Define tolerance to help detect boundary conditions
-    double eps_z =
-      std::min(1e-3 * (outer_radius - inner_radius), 1e-3 * length);
-    double mid_radial_distance = 0.5 * (outer_radius - inner_radius);
+    // First we define the tolerance along the z axis to identify
+    // bottom and top cells.
+    double eps_z = 1e-6 * length;
+
+    // Gather the inner radius from the faces instead of the argument, this is
+    // more robust for some aspect ratios. First initialize the outer to 0 and
+    // the inner to a large value
+    double face_inner_radius = DBL_MAX;
+    double face_outer_radius = 0.;
+
+    // Loop over the cells once to acquire the min and max radius at the face
+    // centers. Otherwise, for some cell ratio, the center of the faces can be
+    // at a radius which is significantly different from the one prescribed.
+    for (const auto &cell : tria.active_cell_iterators())
+      for (const unsigned int f : GeometryInfo<3>::face_indices())
+        {
+          if (!cell->face(f)->at_boundary())
+            continue;
```

Comment:
@kronbichler usually chastizes me about these suggestions, but I typically think that this is easier to write with a positive condition, and no nonlocal exits from the loop iteration:
```suggestion
    for (const auto &cell : tria.active_cell_iterators())
      for (const unsigned int f : GeometryInfo<3>::face_indices())
          if (cell->face(f)->at_boundary())
          {
```